modules/primary_monitoring/module.py

```diff
@@ -38,7 +41,74 @@ def run(self):
         time.sleep(int(sec) if sec is not None else 0)
 
         while True:
-            time.sleep(1)
+            if self.vcm[self.id][PRIM] != self.get_current_view(self.id):
+                self.clean_state()
+
+            self.vcm[self.id][PRIM] = self.get_current_view(self.id)
+
+            # TODO algo4 running in own thread? Talk through resolver
+            # algo 4 running in same thread, have an instance of it
+
+            # self.vcm[self.id][NEED_CHANGE] = ALGO4.suspected()
+
+            if self.resolver.execute(
+               Module.VIEW_ESTABLISHMENT_MODULE, Function.ALLOW_SERVICE):
+                # Line 9-10
+                if (self.vcm[self.id][PRIM] ==
+                   self.get_current_view(self.id) and
+                   self.vcm[self.id][V_STATUS] != enums.V_CHANGE):
+                    self.update_need_chg_set()
+                    # Line 11
+                    if(self.number_of_processor_in_no_service() <
+                       (2 * self.number_of_byzantine + 1)):
+                        self.rep[V_STATUS] = enums.OK
+                    # Line 12
+                    if(self.rep[V_STATUS] == enums.OK and
+                       self.sup_change(3 * self.number_of_byzantine + 1)):
+                        self.rep[V_STATUS] = enums.NO_SERVICE
+                    # Line 13
+                    elif self.sup_change(4 * self.number_of_byzantine + 1):
+                        self.rep[V_STATUS] = enums.V_CHANGE
+                        self.resolver.execute(
+                            Module.VIEW_ESTABLISHMENT_MODULE,
+                            Function.VIEW_CHANGE)
+                # Line 14
+                elif(self.vcm[self.id][PRIM] ==
+                     self.get_current_view(self.id) and
+                     self.vcm[self.id][V_STATUS] == enums.V_CHANGE):
+                    self.resolver.execute(
+                            Module.VIEW_ESTABLISHMENT_MODULE,
+                            Function.VIEW_CHANGE)
+                # Line 15
+                else:
+                    self.clean_state()
+
+            # Send vcm to all nodes
+            self.send_msg()
+
+            # Stopping the while loop, used for testing purpose
+            if(not self.run_forever):
+                break
+
+    # Help functions for run-method
+    def number_of_processor_in_no_service(self):
```

Comment:
Nit: Should be `number_of_processors_in_no_service`

Reply:
Nice!